File: src/building/Building.js
/**
 * Building Class
 * Represents a building or structure in the game
 */
class Building {
    static TYPES = {
        HOUSE: {
            name: 'House',
            width: 1,
            height: 1,
            buildHeight: 40,
            topColor: '#8b4513',
            leftColor: '#654321',
            rightColor: '#7a3d0f',
            sprite: 'house' // Use house sprite
        },
        TOWER: {
            name: 'Tower',
            width: 1,
            height: 1,
            buildHeight: 60,
            topColor: '#696969',
            leftColor: '#505050',
            rightColor: '#5a5a5a',
            sprite: 'building_iso' // Use iso building sprite
        },
        WAREHOUSE: {
            name: 'Warehouse',
            width: 2,
            height: 2,
            buildHeight: 30,
            topColor: '#a0522d',
            leftColor: '#8b4513',
            rightColor: '#964b1a',
            sprite: 'treehouse' // Use treehouse sprite for warehouse
        },
        // Structure types
        WALL_WOOD: {
            name: 'Wooden Wall',
            width: 1,
            height: 1,
            buildHeight: 48,
            structureType: 'wall',
            material: 'wood',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 0,
            walkable: false,
            blocksVision: true,
            health: 100
        },
        WALL_STONE: {
            name: 'Stone Wall',
            width: 1,
            height: 1,
            buildHeight: 48,
            structureType: 'wall',
            material: 'stone',
            spriteSheet: 'walls_rubble',
            spriteIndex: 0,
            walkable: false,
            blocksVision: true,
            health: 200
        },
        WALL_BRICK: {
            name: 'Brick Wall',
            width: 1,
            height: 1,
            buildHeight: 48,
            structureType: 'wall',
            material: 'brick',
            spriteSheet: 'walls_rubble',
            spriteIndex: 8,
            walkable: false,
            blocksVision: true,
            health: 150
        },
        DOOR_WOOD: {
            name: 'Wooden Door',
            width: 1,
            height: 1,
            buildHeight: 48,
            structureType: 'door',
            material: 'wood',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 3,
            walkable: true,
            blocksVision: false,
            canOpen: true,
            health: 80
        },
        WINDOW_GLASS: {
            name: 'Glass Window',
            width: 1,
            height: 1,
            buildHeight: 32,
            structureType: 'window',
            material: 'glass',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 6,
            walkable: false,
            blocksVision: false,
            health: 30
        },
        STAIRS_WOOD: {
            name: 'Wooden Stairs',
            width: 1,
            height: 1,
            buildHeight: 32,
            structureType: 'stairs',
            material: 'wood',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 8,
            walkable: true,
            blocksVision: false,
            connectsFloors: true,
            health: 100
        },
        LADDER_WOOD: {
            name: 'Wooden Ladder',
            width: 1,
            height: 1,
            buildHeight: 48,
            structureType: 'ladder',
            material: 'wood',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 9,
            walkable: true,
            blocksVision: false,
            climbable: true,
            connectsFloors: true,
            health: 80
        },
        HATCH_WOOD: {
            name: 'Wooden Hatch',
            width: 1,
            height: 1,
            buildHeight: 8,
            structureType: 'hatch',
            material: 'wood',
            spriteSheet: 'greenlands_structures',
            spriteIndex: 3,  // Reuses door sprite as hatches are similar to horizontal doors
            walkable: true,
            blocksVision: false,
            canOpen: true,
            connectsFloors: true,
            health: 60
        }
    };
    
    constructor(x, y, type = Building.TYPES.HOUSE, assetLoader = null) {
        this.x = x;
        this.y = y;
        this.type = type;
        this.constructed = true;
        this.health = 100;
        this.assetLoader = assetLoader;
    }
    
    /**
     * Render building
     */
    render(renderer, camera, isometricRenderer, tileScreenPos) {
        if (!this.constructed) return;
        
        // Check if this is a structure type (wall, door, window, etc.)
        if (this.type.spriteSheet && this.type.spriteIndex !== undefined) {
            // Use sprite sheet rendering for structures
            if (this.assetLoader) {
                const spriteSheet = this.assetLoader.getSpriteSheet(this.type.spriteSheet);
                if (spriteSheet) {
                    const sprite = spriteSheet.getSprite(this.type.spriteIndex);
                    if (sprite) {
                        const spriteX = tileScreenPos.x - sprite.width / 2 - camera.x;
                        const spriteY = tileScreenPos.y - sprite.height + 16 - camera.y;
                        
                        renderer.ctx.drawImage(
                            sprite.image,
                            sprite.sx,
                            sprite.sy,
                            sprite.width,
                            sprite.height,
                            spriteX,
                            spriteY,
                            sprite.width,
                            sprite.height
                        );
                        return;
                    }
                }
            }
        }
        
        // Try to use sprite if available (for traditional buildings)
        let buildingSprite = null;
        if (this.assetLoader && this.type.sprite) {
            buildingSprite = this.assetLoader.getImage(this.type.sprite);
        }
        
        if (buildingSprite) {
            // Draw building sprite centered on tile
            const GROUND_ALIGNMENT_OFFSET = 16; // Offset to align sprite base with ground
            const spriteX = tileScreenPos.x - buildingSprite.width / 2 - camera.x;
            const spriteY = tileScreenPos.y - buildingSprite.height + GROUND_ALIGNMENT_OFFSET - camera.y;
            renderer.drawImage(
                buildingSprite,
                spriteX,
                spriteY,
                buildingSprite.width,
                buildingSprite.height
            );
        } else {
            // Fallback to isometric cube rendering
            isometricRenderer.drawIsometricCube(
                tileScreenPos.x,
                tileScreenPos.y,
                64, // tileWidth
                32, // tileHeight
                this.type.buildHeight,
                this.type.topColor,
                this.type.leftColor,
                this.type.rightColor,
                camera
            );
        }
    }
    
    /**
     * Get building info
     */
    getInfo() {
        return {
            name: this.type.name,
            position: { x: this.x, y: this.y },
            health: this.health
        };
    }
}

File: cpp/src/building/Building.cpp
#include "building/Building.h"

Building::Building(int x, int y, BuildingType type)
    : x(x)
    , y(y)
    , type(type)
{
    initializeFromType();
}

std::string Building::getTypeName() const {
    switch (type) {
        case BuildingType::HOUSE:        return "House";
        case BuildingType::TOWER:        return "Tower";
        case BuildingType::WAREHOUSE:    return "Warehouse";
        case BuildingType::WALL_WOOD:    return "Wooden Wall";
        case BuildingType::WALL_STONE:   return "Stone Wall";
        case BuildingType::WALL_BRICK:   return "Brick Wall";
        case BuildingType::DOOR_WOOD:    return "Wooden Door";
        case BuildingType::WINDOW_GLASS: return "Glass Window";
        case BuildingType::STAIRS_WOOD:  return "Wooden Stairs";
        case BuildingType::LADDER_WOOD:  return "Wooden Ladder";
        case BuildingType::HATCH_WOOD:   return "Wooden Hatch";
        default:                         return "Unknown";
    }
}

void Building::initializeFromType() {
    // Default values
    walkable = false;
    blocksVision_ = true;
    structureType = StructureType::NONE;
    
    switch (type) {
        case BuildingType::HOUSE:
            width = 2;
            height = 2;
            buildHeight = 40.0f;
            topColor = glm::vec4(0.8f, 0.2f, 0.2f, 1.0f);
            leftColor = glm::vec4(0.6f, 0.15f, 0.15f, 1.0f);
            rightColor = glm::vec4(0.7f, 0.17f, 0.17f, 1.0f);
            health = 200;
            break;
            
        case BuildingType::TOWER:
            width = 1;
            height = 1;
            buildHeight = 80.0f;
            topColor = glm::vec4(0.5f, 0.5f, 0.5f, 1.0f);
            leftColor = glm::vec4(0.3f, 0.3f, 0.3f, 1.0f);
            rightColor = glm::vec4(0.4f, 0.4f, 0.4f, 1.0f);
            health = 300;
            break;
            
        case BuildingType::WAREHOUSE:
            width = 3;
            height = 3;
            buildHeight = 30.0f;
            topColor = glm::vec4(0.6f, 0.4f, 0.2f, 1.0f);
            leftColor = glm::vec4(0.45f, 0.3f, 0.15f, 1.0f);
            rightColor = glm::vec4(0.52f, 0.35f, 0.17f, 1.0f);
            health = 250;
            break;
            
        // Wall structures
        case BuildingType::WALL_WOOD:
            structureType = StructureType::WALL;
            width = 1;
            height = 1;
            buildHeight = 48.0f;
            topColor = glm::vec4(0.6f, 0.4f, 0.2f, 1.0f);
            leftColor = glm::vec4(0.5f, 0.3f, 0.15f, 1.0f);
            rightColor = glm::vec4(0.55f, 0.35f, 0.17f, 1.0f);
            walkable = false;
            blocksVision_ = true;
            health = 100;
            break;
            
        case BuildingType::WALL_STONE:
            structureType = StructureType::WALL;
            width = 1;
            height = 1;
            buildHeight = 48.0f;
            topColor = glm::vec4(0.5f, 0.5f, 0.5f, 1.0f);
            leftColor = glm::vec4(0.4f, 0.4f, 0.4f, 1.0f);
            rightColor = glm::vec4(0.45f, 0.45f, 0.45f, 1.0f);
            walkable = false;
            blocksVision_ = true;
            health = 200;
            break;
            
        case BuildingType::WALL_BRICK:
            structureType = StructureType::WALL;
            width = 1;
            height = 1;
            buildHeight = 48.0f;
            topColor = glm::vec4(0.7f, 0.3f, 0.2f, 1.0f);
            leftColor = glm::vec4(0.6f, 0.25f, 0.15f, 1.0f);
            rightColor = glm::vec4(0.65f, 0.27f, 0.17f, 1.0f);
            walkable = false;
            blocksVision_ = true;
            health = 150;
            break;
            
        // Door structure
        case BuildingType::DOOR_WOOD:
            structureType = StructureType::DOOR;
            width = 1;
            height = 1;
            buildHeight = 48.0f;
            topColor = glm::vec4(0.5f, 0.35f, 0.2f, 1.0f);
            leftColor = glm::vec4(0.4f, 0.28f, 0.16f, 1.0f);
            rightColor = glm::vec4(0.45f, 0.31f, 0.18f, 1.0f);
            walkable = true;
            blocksVision_ = false;
            health = 80;
            break;
            
        // Window structure
        case BuildingType::WINDOW_GLASS:
            structureType = StructureType::WINDOW;
            width = 1;
            height = 1;
            buildHeight = 32.0f;
            topColor = glm::vec4(0.7f, 0.8f, 0.9f, 0.6f);
            leftColor = glm::vec4(0.6f, 0.7f, 0.8f, 0.6f);
            rightColor = glm::vec4(0.65f, 0.75f, 0.85f, 0.6f);
            walkable = false;
            blocksVision_ = false;
            health = 30;
            break;
            
        // Stairs structure
        case BuildingType::STAIRS_WOOD:
            structureType = StructureType::STAIRS;
            width = 1;
            height = 1;
            buildHeight = 32.0f;
            topColor = glm::vec4(0.55f, 0.4f, 0.25f, 1.0f);
            leftColor = glm::vec4(0.45f, 0.32f, 0.2f, 1.0f);
            rightColor = glm::vec4(0.5f, 0.36f, 0.22f, 1.0f);
            walkable = true;
            blocksVision_ = false;
            health = 100;
            break;
            
        // Ladder structure
        case BuildingType::LADDER_WOOD:
            structureType = StructureType::LADDER;
            width = 1;
            height = 1;
            buildHeight = 48.0f;
            topColor = glm::vec4(0.6f, 0.45f, 0.3f, 1.0f);
            leftColor = glm::vec4(0.5f, 0.37f, 0.24f, 1.0f);
            rightColor = glm::vec4(0.55f, 0.41f, 0.27f, 1.0f);
            walkable = true;
            blocksVision_ = false;
            health = 80;
            break;
            
        // Hatch structure
        case BuildingType::HATCH_WOOD:
            structureType = StructureType::HATCH;
            width = 1;
            height = 1;
            buildHeight = 8.0f;
            topColor = glm::vec4(0.5f, 0.4f, 0.25f, 1.0f);
            leftColor = glm::vec4(0.4f, 0.32f, 0.2f, 1.0f);
            rightColor = glm::vec4(0.45f, 0.36f, 0.22f, 1.0f);
            walkable = true;
            blocksVision_ = false;
            health = 60;
            break;
    }
}
